File: stone_processor/src/main/java/com/github/klee0kai/stone/utils/RecursiveDetector.java
package com.github.klee0kai.stone.utils;

import java.util.LinkedList;
import java.util.Objects;

/**
 * A tortoise is chasing the hare.
 * If the sequence is looped, then the hare will meet the tortoise on the next circle
 *
 * @param <T> sequence type
 */
public class RecursiveDetector<T> {

    private final LinkedList<T> race = new LinkedList<>();
    private long hareStep = 0;


    /**
     * Do next step.
     *
     * @param next next item in sequence
     * @return true if sequence is cycled
     */
    public boolean next(T next) {
        race.add(next);
        if (hareStep++ % 2 == 0) {
            race.pollFirst();
        }
        if (race.size() <= 2) return false;

        T hare = next;
        T turtle = race.getFirst();
        return Objects.equals(hare, turtle);
    }


}

File: stone_ksp/src/commonMain/kotlin/com/github/klee0kai/thekey/stone/ksp/utils/RecursiveDetector.kt
package com.github.klee0kai.thekey.stone.ksp.utils

import java.util.*

/**
 * A tortoise is chasing the hare.
 * If the sequence is looped, then the hare will meet the tortoise on the next circle
 *
 * @param <T> sequence type
</T> */
class RecursiveDetector<T> {

    private val race = LinkedList<T?>()
    private var hareStep: Long = 0


    /**
     * Do next step.
     *
     * @param next next item in sequence
     * @return true if sequence is cycled
     */
    fun next(next: T?): Boolean {
        race.add(next)
        if (hareStep++ % 2 == 0L) {
            race.pollFirst()
        }
        if (race.size <= 2) return false

        val hare = next
        val turtle = race.first()
        return hare == turtle
    }

}
